### PR TITLE
speed up torch.sparse_mask() cpu kernel

### PR DESCRIPTION
**Pull request by weiyangfb**
- `sparse_mask(D, S)` is useful to implement backward for `sparse_addmm()`
- previous `sparse_mask(D, S)` cpu kernel is not parallelized
- this PR speed up the cpu kernel for two separated cases:
  - `D.dim == S.sparse_dim`: simply parallelize the kernel
  - `D.dim > S.sparse_dim`: simply use CUDA kernel implementation
- performance:

`D.dim == S.sparse_dim`
```
>>> nnz = 100000
>>> dims = [1000, 1000]
>>> I = torch.cat([torch.randint(0, dims[0], size=(nnz,)), 
               torch.randint(0, dims[1], size=(nnz,))], 0).reshape(2, nnz)
>>> V = torch.randn(nnz)
>>> size = torch.Size(dims)

>>> S = torch.sparse_coo_tensor(I, V, size).coalesce()
>>> D = torch.randn(dims)

>>> %timeit D.sparse_mask(S)

======= before change =======
6.4 ms ± 684 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

======= after change =======
333 µs ± 89.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

`D.dim > S.sparse_dim`
```
>>> nnz = 100000
>>> dims = [1000, 1000, 2, 2]
>>> I = torch.cat([torch.randint(0, dims[0], size=(nnz,)), 
               torch.randint(0, dims[1], size=(nnz,))], 0).reshape(2, nnz)
>>> V = torch.randn(nnz, dims[2], dims[3])
>>> size = torch.Size(dims)

>>> S = torch.sparse_coo_tensor(I, V, size).coalesce()
>>> D = torch.randn(dims)
%timeit D.sparse_mask(S)

======= before change =======
495 ms ± 41.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

======= after change =======
594 µs ± 68.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 1]
```diff
@@ -394,35 +423,32 @@ SparseTensor& sparse_mask_out_cpu(SparseTensor& r, const Tensor& t, const Sparse
...
-    for (int64_t i = 0; i < r_nnz; i++) {
-      Tensor srcBuffer = t;
-      for (int64_t d = 0; d < sparse_dim; d++) {
-        srcBuffer = srcBuffer.select(0, mask_indices_accessor[d][i]);
-      }
-      Tensor dstBuffer = r_values.select(0, i);
-      dstBuffer.copy_(srcBuffer);
+    LongTensor indices = at::zeros({mask._nnz()}, mask_indices.options());
+
+    for (int64_t d = 0; d < mask.sparse_dim(); d++) {
```
Comment (SsnL) [inline aten/src/ATen/native/sparse/SparseTensor.cpp]: This loop could probably be replaced by https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/SparseTensorUtils.h#L84

Comment (weiyangfb) [reply aten/src/ATen/native/sparse/SparseTensor.cpp]: Thanks! Indeed this can be replaced by `flatten_indices()`, but for some reason it gets slower after the replacement (1.38 ms ± 179 µs after replacement vs 594 µs ± 68.9 µs before it). I will keep the current implementation and note down what the current code is trying to do

**Comment by weiyangfb** [comment]
how can I further improve this PR? cc @SsnL @ezyang

File: torch/_tensor_docs.py [hunk 2]
```diff
@@ -1170,6 +1170,45 @@ def add_docstr_all(method, docstr):
 See :func:`torch.index_select`
 """)
 
+add_docstr_all('sparse_mask',
+               r"""
+sparse_mask(input, mask) -> Tensor
+
+Returns a new SparseTensor with values from Tensor :attr:`input` filtered
+by indices of SparseTensor :attr:`mask`. :attr:`input` and :attr:`mask`
+must have the same shape.
```
Comment (ezyang) [inline torch/_tensor_docs.py]: nit: You could be even more explicit here if you said, values of mask are ignored.

**Comment by ezyang** [comment]
> D.dim > S.sparse_dim: simply use CUDA kernel implementation

I'm not sure what you mean by this.

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 3]
```diff
@@ -394,35 +424,34 @@ SparseTensor& sparse_mask_out_cpu(SparseTensor& r, const Tensor& t, const Sparse
...
-      }
-      Tensor dstBuffer = r_values.select(0, i);
-      dstBuffer.copy_(srcBuffer);
+
+    // Get a flattened sparse indices, similar to NOTE [ Flatten Sparse Indices ].
+    // Keeping this implementation because it is faster than flatten_indices()
+    LongTensor indices = at::zeros({mask._nnz()}, mask_indices.options());
+    for (int64_t d = 0; d < mask.sparse_dim(); d++) {
+      indices.mul_(mask.size(d));
+      // This used to use a buffer but I deoptimized it
```
Comment (ezyang) [inline aten/src/ATen/native/sparse/SparseTensor.cpp]: This comment is out of date now

**Review by ezyang — APPROVED** [review]
Good work, thank you.

File: torch/_tensor_docs.py (hunk 2 above)
Comment (weiyangfb) [reply torch/_tensor_docs.py]: Do you mean: `values` of `mask` SparseTensor is not used? Surely yes I can include that.

**Comment by weiyangfb** [comment]
@ezyang I meant I for code in condition: D.dim > S.sparse_dim, I just copy & paste from https://github.com/pytorch/pytorch/blob/482b1366e68cd32d3ae7671e7ae12c9488464a0a/aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cpp#L33-L51. I feel bad about duplicating code. Let me know if this is not ok.

File: aten/src/ATen/native/sparse/SparseTensor.cpp (hunk 1 above)
Comment (SsnL) [reply aten/src/ATen/native/sparse/SparseTensor.cpp]: Interesting. Even for CUDA? If it is always faster, we should replace `flatten_indices` with this.

Comment (SsnL) [reply aten/src/ATen/native/sparse/SparseTensor.cpp]: not needed for this PR tho
